### PR TITLE
kube-router as default CNI

### PR DESCRIPTION
**Pull request by jnummelin**
Signed-off-by: Jussi Nummelin <jnummelin@mirantis.com>

**Issue**
Fixes xyz

**What this PR Includes**
kube-router as alternative CNI. Why:
- Good alternative with non-overlay networking use cases
- Very versatile, we can use this to replace kube-proxy too later on
- Low(er) resource usage
- Support arm32 arch

The daemonset is using k0sproject/cni-node as the "installer" to put CNI bins into the host level.

**TODOs**

- [x] docs
- [x] make images configurable

### CODE REVIEW

**Review by ncopa — APPROVED** [review]
LGTM

**Comment by jnummelin** [comment]
Ran `sig-network` tests from the conformance suite on LB + 3 ctlr + 3 worker setup with kube-router configured:
```
root@k0s-controller-1:~# cat /etc/k0s/k0s.yaml 
# generated-by-k0sctl 2021-03-30T20:50:58Z
apiVersion: k0s.k0sproject.io/v1beta1
kind: Cluster
metadata:
  name: k0s
spec:
...
  network:
    kuberouter: {}
    provider: kuberouter
```
Results:
```
Plugin: e2e
Status: passed
Total: 5667
Passed: 32
Failed: 0
Skipped: 5635
```
🎉

**Review by jasmingacic — APPROVED** [review]

**Review by jasmingacic — DISMISSED** [review]

File: docs/configuration.md [hunk 1]
```diff
@@ -99,7 +99,7 @@ Using type `etcd` will make k0s to create and manage an elastic etcd cluster wit
 
 ### `spec.network`
 
-- `provider`: Network provider, either `calico` or `custom`. In case of `custom` user can push any network provider.
+- `provider`: Network provider, either `calico`, `kuberouter` or `custom`. In case of `custom` user can push any network provider. (default `calico`)
```
Comment (trawler) [inline docs/configuration.md:102]: default shouldn't be cube-router?

Comment (jnummelin) [reply docs/configuration.md:102]: 👍  good catch, default indeed after this PR is merged is kuberouter. I'll fix

File: docs/networking.md [hunk 2]
```diff
@@ -2,11 +2,16 @@
 
 ## In-cluster networking
 
-k0s uses Calico as the default, built-in network provider. Calico is a container networking solution making use of layer 3 to route packets to pods. It supports for example pod specific network policies helping to secure kubernetes clusters in demanding use cases. Calico uses vxlan overlay network by default. Also ipip (IP-in-IP) is supported by configuration.
+k0s support two CNI providers out-of-box plus the ability to bring your own CNI config.
+
+k0s uses [Calico](https://www.projectcalico.org/) as the default, built-in network provider. Calico is a container networking solution making use of layer 3 to route packets to pods. It supports for example pod specific network policies helping to secure kubernetes clusters in demanding use cases. Calico uses vxlan overlay network by default. Also ipip (IP-in-IP) is supported by configuration.
```
Comment (trawler) [inline docs/networking.md:7]: this part of the docs should also be updated

File: docs/networking.md [hunk 3]
```diff
@@ -2,12 +2,36 @@
 
 ## In-cluster networking
 
-k0s uses Calico as the default, built-in network provider. Calico is a container networking solution making use of layer 3 to route packets to pods. It supports for example pod specific network policies helping to secure kubernetes clusters in demanding use cases. Calico uses vxlan overlay network by default. Also ipip (IP-in-IP) is supported by configuration.
+k0s support two CNI providers out-of-box plus the ability to bring your own CNI config.
```
Comment (mviitane) [inline docs/networking.md:5]: support->supports

File: docs/networking.md [hunk 4]
```diff
@@ -2,12 +2,36 @@
...
+**Calico:**
+- does NOT support armv7
+- uses bit more resources
+- does support dual-stack (IPv4/IPv6) networking
+- does support Windows nodes
 
+**Note:** After the cluster has been initialized with one network provider, it is not currently supported to change the network provider. Changing network provider requires full re-deployment of the cluster.
 ## Controller(s) - Worker communication
 
 As one of the goals of k0s is to allow deployment of totally isolated control plane we cannot rely on the fact that there is an IP route between controller nodes and the pod overlay network. To enable this communication path, which is mandated by conformance tests, we use [Konnectivity service](https://kubernetes.io/docs/tasks/extend-kubernetes/setup-konnectivity/) to proxy the traffic from API server (control plane) into the worker nodes. Possible firewalls should be configured with outbound access so that Konnectivity agents running on the worker nodes can establish the connection. This ensures that we can always fulfill all the Kubernetes API functionalities, but still operate the control plane in total isolation from the workers.
```
Comment (mviitane) [inline docs/networking.md:37]: "overlay" word should be removed?

**Review by mviitane — DISMISSED** [review]
Looks good. Only 2 very minor comments.

**Comment by trawler** [comment]
just a nitpick but my ocd won't let it go ;) can we add a fix to the typo in the "custom pors" smoke test? https://github.com/k0sproject/k0s/blob/dcb0e0dcb674234637e7366342f5bc02994dad23/.github/workflows/go.yml#L115

**Review by trawler — APPROVED** [review]

**Pull request merged**